sklearn/metrics/ranking.py

```diff
@@ -258,7 +258,7 @@ def _binary_roc_auc_score(y_true, y_score, sample_weight=None):
 
         fpr, tpr, tresholds = roc_curve(y_true, y_score,
                                         sample_weight=sample_weight)
-        return auc(fpr, tpr, reorder=True)
+        return auc(fpr, tpr, reorder=False)
```

Comment:
This change is quite bold and it doesn't seem to me like you tried to estimate how much of an effect this will have. Passing the tests is one thing but I am afraid that it's not enough ...

If I look at the code (in master), it looks like reorder=True is only used in _binary_roc_auc_score so I would be enclined to think it is there for a reason. Look at git blame and you'll find hopefully a good commit message or a related PR/issue to understand the reason of the change.

Reply:
@lesteve Thanks,
The code is introduced in https://github.com/scikit-learn/scikit-learn/commit/9f18586ebfc835810afff50fcc3449265501bbcb. Since then, the implementation of roc_curve has been changed. fpr and tpr is already guaranteed to be increasing now.

Reply:
> fpr and tpr is already guaranteed to be increasing now.

I don't think that's true, the tests fail if I use reorder=True.

Reply:
@lesteve Thanks.
This is exactly the case. Do not sort fpr and tpr can not only save time, but also improve stability.
Here is a detailed explanation:
![2017-09-20 170022](https://user-images.githubusercontent.com/12003569/30635554-581a41e0-9e25-11e7-842a-1a0c6960045e.jpg)
Suppose that we get A, B, C on the curve, we get S(AA'B'B) from auc. This is right.
However, if we introduce some error and get D instead of C, if we sort, we get A, D, B. Finally, we will get S(AA'D'D)+S(DD'B'B) from auc, this is why we get pretty big difference.
If we notice that fpr and tpr is already increasing (not considering the minor error introduced by python) and don't sort, we get A, B, D. Finally, we will get S(AA'B'B)-S(BB'D'D) from auc, note that B'D' is really small, so the result is very close to the right result.

Reply:
I do get this on a toy example like yours and it seems like this is indeed what is happening in the failing test. What I am worried about is that you did not evaluate the impact of this change outside of the tests. For example how confident are you that your 1e-10 is good enough for a wide range of input parameters? I would be a lot more reassured if you spent a bit of time looking at this kind of things.

Either generate random `y_true`, `y_score` and look at the distribution of `np.diff(tpr)`, `np.diff(fpr)` returned by roc_curve, or run a variety of models on some classification task, or think of something better.

To be honest it feels like roc_curve should live up to its docstring of returning increasing fpr and tpr but if I understand https://github.com/scikit-learn/scikit-learn/pull/9786#issuecomment-330382029 correctly @jnothman was saying there has been attempts on and this issue and that they failed.

Reply:
@lesteve Thanks. I can first give a simple analysis to see if this satisfied you.
When we set the tolerance to 1e-10, according to the calculation method of np.trapz, even if we fail to sort x to the correct order in extreme situations (e.g., users provide x1-x2<1e-5 and weight1-weight2<1e-5), the difference for each point we get will be <2e-10*max(y). From my perspective, if users are not sure whether their data should be sorted or not, they should set reorder=true instead of relying on the judgement from scikit-learn and they will get very very precise result. If their data is already sorted, than the tolerance in the judgement will not affect the result at all. On the contrary, it will improve the result as shown in this PR.